refactor: remove remaining unused code from cpp-subprocess

theStack opened the pull request:
This PR removes remaining code that is unused within the cpp-subprocess module (templates and commented out code). Happy to add more removals if anyone finds more unused parts. Note that there are some API functions of the `Popen` class that we don't use, e.g. `wait()`, `pid()`, `poll()`, `kill()`, but they sound IMHO common enough to be useful in the future, so not sure how deep we should go there.

theStack committed:
remove unused templates from cpp-subprocess

The templates `is_ready`, `param_pack`, and `has_type` are not used
anywhere, so let's remove them.

theStack committed:
remove commented out code in cpp-subprocess

fanquake commented:
> Note that there are some API functions of the Popen class that we don't use, e.g. wait(), pid(), poll(), kill(), but they sound IMHO common enough to be useful in the future, so not sure how deep we should go there.

Deleting as much as possible seems fine. If code turns out to be needed in future (although I wouldn't think we are going to expand this interface), it can be retrieved from the git history.

Sjors commented:
As long as #29868 works on top of this, it's fine by me. I'll do some testing.

hebasto reviewed (approved):
ACK 908c51fe4afeba0af500c6275027b1afa1b3bd19. It is compatible with ~https://github.com/bitcoin/bitcoin/pull/29961~  #29868.

I'd suggest to remove the mentioning of the `check_output` function from all comments. That function was not even pulled from the upstream.

hebasto commented:
> > Note that there are some API functions of the Popen class that we don't use, e.g. wait(), pid(), poll(), kill(), but they sound IMHO common enough to be useful in the future, so not sure how deep we should go there.
> 
> Deleting as much as possible seems fine. If code turns out to be needed in future (although I wouldn't think we are going to expand this interface), it can be retrieved from the git history.

I agree.

theStack committed:
remove unused method `Popen::kill` from cpp-subprocess

theStack committed:
update comments in cpp-subprocess (check_output references)

Remove obsolete `check_output` references in the comments and remove
the numbering of the Popen API methods, as they don't seem to provide a
value and just make diffs larger for future changes.

fjahr commented on src/util/subprocess.h:
```diff
@@ -904,8 +904,6 @@ class Streams
  * 5. pid()              - PID of the spawned child.
- * 6. poll()             - Check the status of the running child.
- * 7. kill(sig_num)      - Kill the child. SIGTERM used by default.
  * 8. send(...)          - Send input to the input channel of the child.
```
The numbering of this list is off after this change. That might be confusing to someone in the future. I would suggest either updating the following numbering or something like.

```
 * 6. (deleted)
 * 7. (deleted)
```

fjahr replied:
Ok, strange, I guess I was looking at an outdated version of the code or something.

hebasto commented on src/util/subprocess.h:
```diff
@@ -1009,27 +955,12 @@ class Popen
 
   int wait() noexcept(false);
 
-  int poll() noexcept(false);
```
ed39d261dc056e700b79da769e085c8af389451d:

The `Popen::poll()` function is used in #29868. I hope that an improved/fixed Windows implementation of the `Popen::retcode()`, which I don't have now, will allow to drop it.

fjahr commented:
> ACK [908c51f](https://github.com/bitcoin/bitcoin/commit/908c51fe4afeba0af500c6275027b1afa1b3bd19). It is compatible with #29961.

@hebasto #29961 is this PR here, did you mean #29868 there?

fjahr commented:
Code review ACK 8b52e7f628304e83b0e36fd97e617de0f71c5a62

hebasto reviewed (approved):
ACK 8b52e7f628304e83b0e36fd97e617de0f71c5a62.

achow101 commented:
ACK 8b52e7f628304e83b0e36fd97e617de0f71c5a62
